src/QsCompiler/QirGeneration/DebugInfoManager.cs

```diff
@@ -170,46 +393,138 @@ private DebugInfoBuilder GetOrCreateDIBuilder(string sourcePath)
         }
 
         /// <summary>
-        /// If DebugFlag is set to true, creates Adds top level debug info to the module,
-        /// Note: because this is called from within the constructor of the GenerationContext,
-        /// we cannot access this.Module or anything else that uses this.sharedState
+        /// Creates a debug type for a particular Q# type. The debug type connects a debug info type with its
+        /// corresponding LLVM Native type.
         /// </summary>
-        internal void AddTopLevelDebugInfo(BitcodeModule owningModule, ImmutableArray<QsQualifiedName> entryPoints)
+        /// <param name="resolvedType">The Q# type to create a DebugType from.</param>
+        /// <param name="dIBuilder">The <see cref="DebugInfoBuilder"/> to use to create the debug type.</param>
+        /// <returns>The Debug Type corresponding to the given Q# type. Returns null if the debug type is not implemented.</returns>
+        private IDebugType<ITypeRef, DIType>? GetDebugTypeFor(ResolvedType resolvedType, DebugInfoBuilder dIBuilder)
```

Comment:
@bettinaheim we'd discussed using a type transformation, but I made this helper function for the following reasons. Let me know if you think I should still set it up as a type transformation.

- I tried to have an abstract types class that NativeLlvmType (the existing Type class) and DebugType both inherit from, but C# doesn't support covariant return types in overrides, so I can't specify that we need outputs that implement IDebugType (rather than ITypeRef which it implements).
- If NativeType and DebugType don't both inherit from some Type, then can't reuse TypeTransformation. It seems like quite the overhead to implement another TypeTransformation when we're not actually transforming anything (the existing TypeTransformation just uses a builtType side effect to get the LLVM Native Type)
- We need a dIBuilder to create a DebugType, so we would need a different TypeTransformation for each source file so that it can have its own dIBuilder.

Reply:
I think if you populate it here to include all types, you will effectively just end up reimplementing the type transformation without using a type transformation. I am hence not sure I understand your second bullet point, and also the third bullet is not quite clear to me. I'd be happy to chat to better understand what you mean.